sw/device/lib/dif/dif_padctrl.h

```diff
@@ -0,0 +1,196 @@
+ */
+dif_pinmux_registers_lock_error_t dif_pinmux_registers_lock(
+    dif_pinmux_t *pinmux);
+
+/**
+ * Padctrl attribute enable routines error codes.
+ */
+typedef enum dif_padctrl_enable_attr_error {
+  kDifPadctrlEnableErrorNone = 0,     /**< Success. */
+  kDifPadctrlEnableErrorNotSupported, /**< Attribute is not supported.  */
+  kDifPadctrlEnableErrorConflict,     /**< Could not enable attribute because
+                                         it conflicts with another already
+                                         enabled attribute. */
+  kDifPadctrlEnableInvalidArgument,   /**< Input paramter is not valid. */
+  kDifPadctrlEnableRegistersLocked,   /**< Registers are locked and cannot be
+                                         modified. */
+} dif_padctrl_enable_error_t;
+
+/**
+ * Enables an attribute for an MIO pad
+ *
+ * Not all pads implement all attributes and some combinations of attributes
+ * cannot be enabled together. |dif_padctrl_mio_enable_attr| will check for
+ * this and return an appropriate error. Note that for additional attributes as
+ * the meaning is implementation defined |dif_padctrl_mio_enable_attr| cannot
+ * check for invalid combinations that involve them.
+ *
+ * @param padctrl Padctrl state data.
+ * @param mio Which MIO pad to enable an attribute for.
+ * @param attr Attribute to enable.
+ * @return dif_padctrl_enable_error_t
+ */
+dif_padctrl_enable_error_t dif_padctrl_mio_enable_attr(
+  const dif_padctrl_t *padctrl,
+  dif_padctrl_mio_t mio,
+  dif_padctrl_attr_t attr);
+
+/**
+ * Enables an attribute for a DIO pad
+ *
+ * Not all pads implement all attributes and some combinations of attributes
+ * cannot be enabled together. |dif_padctrl_dio_enable_attr| will check for
+ * this and return an appropriate error. Note that for additional attributes as
+ * the meaning is implementation defined |dif_padctrl_dio_enable_attr| cannot
+ * check for invalid combinations that involve them.
+ *
+ * @param padctrl Padctrl state data.
+ * @param dio Which DIO pad to enable an attribute for.
+ * @param attr Attribute to enable.
+ * @return dif_padctrl_enable_error_t
+ */
+dif_padctrl_enable_error_t dif_padctrl_dio_enable_attr(
+  const dif_padctrl_t *padctrl,
+  dif_padctrl_dio_t dio,
+  dif_padctrl_attr_t attr);
+
+/**
+ * Padctrl attribute disable routines error codes.
+ */
+typedef enum dif_padctrl_disable_attr_error {
+  kDifPadctrlDisableErrorNone = 0,   /**< Success. */
+  kDifPadctrlDisableInvalidArgument, /**< Input paramter is not valid. */
+  kDifPadctrlDisableRegistersLocked, /**< Registers are locked and cannot be
+                                        modified. */
+} dif_padctrl_disable_attr_error_t;
+
+/**
+ * Disables an attribute for an MIO pad
+ *
+ * @param padctrl Padctrl state data.
+ * @param mio Which MIO pad to disable an attribute for.
+ * @param attr Attribute to disable
+ * @return dif_padctrl_disable_attr_error_t
+ */
+dif_padctrl_disable_attr_error_t dif_padctrl_mio_disable_attr(
```

Comment:
Can't enable and disable routines be combined, and `flag` passed in to instruct whether to enable/disable the attribute for mio?

Or are enable and disable routines semantically different, and operate on different registers?

Reply:
I think combining them is sensible.

Initially I wrote it like this because the enable could fail and the disable couldn't but then I realised there were various failure reasons for disable which end up being a subset of the enable failures.